hs_labels/models.py

```diff
@@ -0,0 +1,876 @@
+
+    This is typically a programming error and should be entered into
+    issue tracker and resolved.
+
+    *Catching this exception is not recommended.*
+    """
+    def __str__(self):
+        return repr("HS Usage Exception: " + self.message)
+
+    pass
+
+
+####################################
+####################################
+# The labeling system allows
+# a) labeling of resources.
+# b) assigning folders to resources.
+# c) marking public resources to be "of interest"
+####################################
+####################################
+
+
+####################################
+# privileges are numeric codes 1-4
+####################################
+class LabelCodes(object):
+    """
+    Label codes describe the kinds of labels that a user can create for a thing:
+
+        * 1 or LabelCodes.LABEL:
+            a regular label for the object.
+
+        * 2 or LabelCodes.FOLDER:
+            a folder for the object.
+
+        * 3 or LabelCodes.FAVORITE:
+            marks the object as a favorite.
+
+        * 4 or LabelCodes.MINE:
+            marks the object to be included in my resources
+
+        * 5 or LabelCodes.SAVEDLABEL:
+            creates a label that can be used later for labeling a resource
+
+    """
+    LABEL = 1
+    FOLDER = 2
+    FAVORITE = 3
+    MINE = 4
+    SAVEDLABEL = 5
+    LABEL_CHOICES = (
+        (LABEL, 'Label'),
+        (FOLDER, 'Folder'),
+        (FAVORITE, 'Favorite'),
+        (MINE, 'Mine'),
+        (SAVEDLABEL, 'Saved Label')
+    )
+
+####################################
+# user access to resources
+####################################
+
+
+class UserResourceLabels(models.Model):
+    """ Labels of a user for a resource
+
+    This model stores labels of an individual user, like an access
+    control list; There are several kinds of labels documented in LabelCodes.
+    These are similar in implementation but differ in semantics.
+    """
+
+    kind = models.IntegerField(choices=LabelCodes.LABEL_CHOICES,
+                               editable=False,
+                               default=LabelCodes.LABEL)
+    # provenance of label start
+    start = models.DateTimeField(editable=False, auto_now=True)
+
+    # if end is non-null, then the label is no longer valid. Not absolutely sure we need this.
+    # end  = models.DateTimeField(editable=False, null=True, auto_now=False)
+
+    # Utilize an underlying class of User to hold labels for resources.
+    # This creates an implicit query User.ulabels that represents all labels
+    # for a user.
+    ulabels = models.ForeignKey('UserLabels', null=False, editable=False,
+                                related_name='ul2url',         # unused but must be defined and unique
+                                help_text='user assigning a label')
+    # There is a basic design question as to whether the target of a label should be a
+    # Resource proper, or an annotation class that contains mainly methods for the resource.
```

Comment:
Yes, except for the mitigating circumstance that the User model is a system model. So, to associate methods with a User, I had to add one level of indirection. That being the only reasonable solution for User, using a parallel structure for resources seemed the easiest to remember (and parallel to hs_access_control conventions, which by now people are familiar with using).

In the future, we may want to use this with Group as well, so that we'll continue to need indirection.